### PR TITLE
Add filter to construction menu

### PR DESCRIPTION
**Pull request by sboigelot**
ezgif com-crop

> Just because I'm blind with these kind of list

### CODE REVIEW

**Comment by vogonistic** [comment]
Does it have focus automatically or does it require a click? I'm asking because I'm hoping someone will implement a DF or Warcraft-like set of short keys, so `c d` would be short for construct door. If the filter doesn't have automatic focus then both methods can coexist.

**Comment by sboigelot** [comment]
@vogonistic no automatic focus for now but it's one extra line of code if we decide to go for it

**Comment by Raketfart** [comment]
Good idea! 
Is there any chance of increasing the scroll speed of the construction menu, while we're at it?

**Comment by sboigelot** [comment]
@Raketfart sure, no prob, done :)

**Comment by bjubes** [comment]
The menu seems to be anchored incorrectly. This is what i'm getting.
capture

**Comment by jaselito** [comment]
Having the same problem as bjubes...

**Comment by jaselito** [comment]
Somehow, the Pos Y in the rect transform of the construction menu prefab got set to -384... getting it back to 0 should solve the problem

**Comment by sboigelot** [comment]
@bjubes @jaselito the anchor position of the construction menu should be fix now.

File: Assets/Scripts/Controllers/KeyboardManager.cs [hunk 1]
```diff
@@ -22,12 +23,15 @@ public KeyboardManager()
...
         instance = this;
         mapping = new Dictionary<string, KeyboadMappedInput>();
+        ModalInputFields = new List<InputField>();
 
         TimeManager.Instance.EveryFrameNotModal += (time) => Update();
 
         ReadXmlOrJsonAfterWeDecide();
     }
 
+    public List<InputField> ModalInputFields { get; set; }
```
Comment (mikejbrown) [inline Assets/Scripts/Controllers/KeyboardManager.cs]: Stylecop complains: `SA1204 : CSharp.Ordering : All static public properties must be placed before all non-static public properties. (Assembly-CSharp)`

File: Assets/Scripts/UI/InGameUI/MenuLeft/ConstructionMenu.cs [hunk 2]
```diff
@@ -50,6 +50,24 @@ public void RebuildMenuButtons(bool showAllFurniture = false)
...
+
+        List<Transform> childs = contentTransform.Cast<Transform>().ToList();
+
+        foreach (Transform child in childs)
+        {
+            Text buttonText = child.gameObject.transform.GetComponentInChildren<Text>();
+
+            string buildableName = buttonText.text;
+
+            bool nameMatchFilter = string.IsNullOrEmpty(filterText) || buildableName.ToLower().Contains(filterText.ToLower());
```
Comment (mikejbrown) [inline Assets/Scripts/UI/InGameUI/MenuLeft/ConstructionMenu.cs]: Matches are not anchored to the start of the name, e.g. if I type "dd" this matches "Ladder". Is this desired behaviour?

Comment (sboigelot) [inline Assets/Scripts/UI/InGameUI/MenuLeft/ConstructionMenu.cs]: Yep this is the desired behavior. I want to see "Cloning pod" if I search pod for instance.

File: Assets/Scripts/Controllers/KeyboardManager.cs (hunk 1 above)
Comment (sboigelot) [inline Assets/Scripts/Controllers/KeyboardManager.cs]: Fixed, thanks!

File: Assets/Scripts/UI/InGameUI/MenuLeft/ConstructionMenu.cs (hunk 2 above)
Comment (mikejbrown) [inline Assets/Scripts/UI/InGameUI/MenuLeft/ConstructionMenu.cs]: Good point :+1:

**Pull request merged**